### ISSUE TITLE
Add option to the web service sender to skip WSDL/XSD validation at dispatch time

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3046]
This is important for servers that don't present valid WSDLs, or that don't present WSDLs at all. Currently that definition is validated as a part of creating the JAX-WS Dispatch, so even at dispatch time it's sending a GET request for every message. We need an option to turn that request and subsequent validation off, so that the underlying endpoint can still be invoked.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3133
Reporter: narupley
Created: 2014-01-30T17:11:16.000-0800

**Comment by rbeckman-nextgen** [comment]
In the IHE (PIX/PDQ/XDS) world, servers all have backends that meet specification, but may not always present proper wsdls.  Some systems present WSDLs, but they are wrong/junk. This seems to be more common with .NET developed servers (it seems that the developers did not 'implement' a wsdl). There is a crazy workaround that allows us to query these types of systems, but it would much preferred if there was a simple check box to fix this problem instead.

Imported Comment. Original Details:
Author: dschramm
Created: 2014-01-30T17:17:25.000-0800

**Comment by rbeckman-nextgen** [comment]
Current workaround:

1) Find a locally running service without SSL or other security on it that you are able to click "Get Operations" with.
2) Copy the wsdl and put it in /opt/mirthconnect/public_html/<SERVICENAME>.wsdl
3) Update the permissions to be mirth:mirth
4) Load a MC channel with a WS Sender destination connector, put in the wsdl url: http://localhost:8080/<SERVICENAME>.wsdl
5) Click Get Operations and allow it to pass.
6) Copy and paste a known good soap envelope.
7) Save and Deploy the channel. Currently the channel is pointing to the WRONG endpoint as per what is in the WSDL hosted locally.
8) vi the file /opt/mirthconnect/public_html/<SERVICENAME>.wsdl and replace the two instances of the good endpoints (in my case: https://207.38.40.58:443/pixpdq/PIXManager_Service) with the endpoints of the system that cant present a wsdl (in my case: http://10.242.100.99:10449/qd.mp.ens.sv.HL7Version3.cls)
9) At this point, DO NOT reopen the channel. Just send a message through. Clicking get operations or other buttons in the channel may require you to restart this process.


Imported Comment. Original Details:
Author: dschramm
Created: 2014-01-30T17:23:27.000-0800

**Comment by rbeckman-nextgen** [comment]
Is it possible to finish this ticket and get the fix to an Appliance release within 6 months?

Imported Comment. Original Details:
Author: dschramm
Created: 2014-02-03T13:19:41.000-0800

**Comment by rbeckman-nextgen** [comment]
Possible, yes...

Imported Comment. Original Details:
Author: narupley
Created: 2014-02-03T13:26:56.000-0800

**Comment by rbeckman-nextgen** [comment]
Ugh, this is turning out to be very complicated, not yet sure if it's even possible with JAX-WS (unless we override and re-implement a lot of com.sun.* internal things).

In your workaround case, you still have a WSDL. It's not pointing to the correct endpoint, but the PIX/PDQ specification is still the same so JAX-WS is still able to externalize the SOAP envelope during dispatch. If we allowed the ability to dynamically set the endpoint address regardless of the WSDL service/binding, then this would solve the issue for you, right?

We have an open issue for this (MIRTH-3431) which was needed for other reasons anyway. It was for some reason moved out of 3.1.0, but can be moved back in.

If the server just doesn't present a WSDL at all and also doesn't follow some known specification that you can emulate with a similar WSDL (even if the endpoints are incorrect), then there's really nothing you can do... JAX-WS needs to know the correct abstract and concrete definition, including the expected message and response schema. If it doesn't know what that is, then it cannot parse a SOAP envelope (or response). We could bake in some hacky special case where if the WSDL isn't supplied then just do a straight HTTP POST, but then we would have to handle MTOM all by ourselves, and we'd lost the ability to correctly validate SOAP faults in the response, etc. At that point, you might as well just be using an HTTP Sender instead.

So to recap, in order solve an issue like this you would need to:
* Get a WSDL from an equivalently specified service, and put it somewhere on the filesystem (or host it somewhere)
* Point to that WSDL in the Web Service Sender settings (file or HTTP URLs both work)
* Do all the normal Get Operations / etc. configuration
* Include the actual endpoint address you want to dispatch to in a new WS Sender field (with MIRTH-3431)

Does that sound like a reasonable compromise?

Imported Comment. Original Details:
Author: narupley
Created: 2014-07-25T12:37:07.000-0700

**Comment by rbeckman-nextgen** [comment]
Sounds good to me

Imported Comment. Original Details:
Author: dschramm
Created: 2014-07-25T13:11:49.000-0700

**Comment by rbeckman-nextgen** [comment]
This problem is now solved due to MIRTH-3431. Instead of having to do a hacky workaround to modify the endpoints within the WSDL itself, you can now override the location URI itself in the Web Service Sender settings. A WSDL with a similar specification (enough that JAX-WS is able to externalize the SOAP envelope and response) still needs to exist though, on the filesystem or otherwise.

Imported Comment. Original Details:
Author: narupley
Created: 2014-07-25T17:35:09.000-0700

**Comment by rbeckman-nextgen** [comment]
Changing the resolution to Won't Fix because the actual feature/improvement described was _**not**_ added. Instead the resolution is a sort of workaround made easy by a separate issue.

Imported Comment. Original Details:
Author: narupley
Created: 2014-09-02T12:50:45.000-0700
